### PR TITLE
Use pipelining for remaining Redis operations

### PR DESCRIPTION
**Pull request by FZambia**
This is related to #138 . This pr contains not only pipelining but also several other stuff:

* fix extra call to removing presence for channels with `presence` option disabled
* benchmarks for Memory and Redis engine operations: publish, publish with history, adding presence, getting history, getting presence.
* new subscribe/usubscribe algorithm to exclude theoretical non-deterministic state of Redis channel subscriptions. 

Now benchmarks results on my notebook.

For Redis Engine:

```
BenchmarkPublish-4                       1000000              2153 ns/op             604 B/op         11 allocs/op
BenchmarkPublishWithHistory-4             200000             11762 ns/op             832 B/op         17 allocs/op
BenchmarkOpAddPresence-4                   10000            111555 ns/op             847 B/op         21 allocs/op
BenchmarkOpAddPresenceParallel-4           20000             63271 ns/op             787 B/op         19 allocs/op
BenchmarkOpPresence-4                      10000            116003 ns/op            1877 B/op         29 allocs/op
BenchmarkOpPresenceParallel-4              20000             66277 ns/op            1737 B/op         24 allocs/op
BenchmarkOpHistory-4                       20000             97170 ns/op            1576 B/op         42 allocs/op
BenchmarkOpHistoryParallel-4               30000             53960 ns/op            1546 B/op         41 allocs/op
```

This is not CPU-bound bench so I artificially increased parallel level setting `b.SetParallelism(12)` and results look like:

```
BenchmarkPublish-4                       1000000              2289 ns/op             609 B/op         11 allocs/op
BenchmarkPublishWithHistory-4             200000             12981 ns/op             837 B/op         17 allocs/op
BenchmarkOpAddPresence-4                   10000            114720 ns/op             879 B/op         22 allocs/op
BenchmarkOpAddPresenceParallel-4          100000             20339 ns/op             746 B/op         16 allocs/op
BenchmarkOpPresence-4                      10000            134971 ns/op            1839 B/op         28 allocs/op
BenchmarkOpPresenceParallel-4              50000             25833 ns/op            1677 B/op         21 allocs/op
BenchmarkOpHistory-4                       10000            115476 ns/op            1679 B/op         44 allocs/op
BenchmarkOpHistoryParallel-4              200000              9938 ns/op            1444 B/op         36 allocs/op
```

And the same for memory engine:

```
BenchmarkPublish-4                       3000000               410 ns/op             112 B/op          2 allocs/op
BenchmarkPublishWithHistory-4            1000000              2543 ns/op            1437 B/op          5 allocs/op
BenchmarkOpAddPresence-4                10000000               184 ns/op               0 B/op          0 allocs/op
BenchmarkOpAddPresenceParallel-4         5000000               313 ns/op               0 B/op          0 allocs/op
BenchmarkOpPresence-4                    1000000              1461 ns/op             944 B/op          2 allocs/op
BenchmarkOpPresenceParallel-4            1000000              1093 ns/op             944 B/op          2 allocs/op
BenchmarkOpHistory-4                    10000000               125 ns/op               0 B/op          0 allocs/op
BenchmarkOpHistoryParallel-4            20000000                81.8 ns/op             0 B/op          0 allocs/op
```

@banks I simplified subscribing algorithm comparing to what I showed you before (there are no `noops` now).

### CODE REVIEW

File: libcentrifugo/engine/engineredis/engine.go [hunk 1]
```diff
@@ -956,53 +960,97 @@ func (e *Shard) runPubSub(numWorkers int) {
...
+					if batch[i].subscribe {
+						batch[i].done(nil)
+					}
+				}
+				if len(unsubChIDs) > 0 {
+					err := conn.Unsubscribe(unsubChIDs...)
+					if err != nil {
+						logger.ERROR.Printf("RedisEngine Unsubscriber error: %v\n", err)
+						// Here we should only resolve with error unsubscribe requests and
+						// noop requests.
```
Comment (banks) [inline libcentrifugo/engine/engineredis/engine.go]: There are no no-op requests any more are there?

Comment (FZambia) [inline libcentrifugo/engine/engineredis/engine.go]: yes, found a way to simplify code

**Comment by FZambia** [comment]
@banks thanks a lot for checking this out!

File: libcentrifugo/engine/engineredis/engine.go (hunk 1 above)
Comment (FZambia) [inline libcentrifugo/engine/engineredis/engine.go]: aaa! you mean wrong comment, yep, will fix it

**Pull request merged**